Bluetooth: Host: Add missing buffer length check

ekleezg opened the pull request:
Modified to check the length of the remaining data in buffer before processing the next report. The length check is missing in the cont routine.

Thalley commented on subsys/bluetooth/host/scan.c:
```diff
@@ -721,6 +721,10 @@ void bt_hci_le_adv_ext_report(struct net_buf *buf)
 		reset_reassembling_advertiser();
 
 cont:
+		if (buf->len < evt->length) {
+			LOG_ERR("Unexpected end of buffer");
+			break;
+		}
```
Doesn't this check fit better right after 
```c
evt = net_buf_pull_mem(buf, sizeof(*evt));
``` 
? 

Otherwise we still attempt to parse the event if it has invalid length.

ekleezg replied:
https://github.com/zephyrproject-rtos/zephyr/blob/b30b74b850ba0b8b696bfed86bf0fcd5d896fbeb/subsys/bluetooth/host/scan.c#L683-L696

@Thalley I believe that above check routine should be positioned right after the pulling of evt. Upon your confirmation, I will change it.

jhedberg replied:
I think the key thing is to do the check at least before line 707. The sooner the better though.

Thalley commented on subsys/bluetooth/host/scan.c:
```diff
@@ -620,6 +620,19 @@ void bt_hci_le_adv_ext_report(struct net_buf *buf)
+		if (evt->length > buf->len) {
+			LOG_WRN("Adv report corrupted (wants %u out of %u)", evt->length, buf->len);
+
+			/* Start discarding irrespective of the `more_to_come` flag. We
+			 * assume we may have lost a partial adv report in the truncated
+			 * data.
+			 */
+			reassembling_advertiser.state = FRAG_ADV_DISCARDING;
+			net_buf_reset(buf);
```
Would it make sense to at least check the `if (evt_type & BT_HCI_LE_ADV_EVT_TYPE_LEGACY)` before discarding any partial extended advertisements?

ekleezg committed:
Bluetooth: Host: Add missing buffer length check

Modified to check the length of the remaining data in buffer
before processing the next report. The length check is missing
in the cont routine.

Signed-off-by: Eunkyu Lee <mochaccino.00.00@gmail.com>

Thalley commented on subsys/bluetooth/host/scan.c:
```diff
@@ -620,6 +620,24 @@ void bt_hci_le_adv_ext_report(struct net_buf *buf)
+
+			net_buf_reset(buf);
+
+			if (evt_type & BT_HCI_LE_ADV_EVT_TYPE_LEGACY) {
+				return;
+			}
+
+			/* Start discarding irrespective of the `more_to_come` flag. We
+			 * assume we may have lost a partial adv report in the truncated
+			 * data.
+			 */
+			reassembling_advertiser.state = FRAG_ADV_DISCARDING;
+
+			return;
```
```suggestion
			if ((evt_type & BT_HCI_LE_ADV_EVT_TYPE_LEGACY) == 0) {
				/* Start discarding irrespective of the `more_to_come` flag. We
				 * assume we may have lost a partial adv report in the truncated
				 * data.
				 */
				reassembling_advertiser.state = FRAG_ADV_DISCARDING;
			}

			return;
```

This would reduce the number of `return` statements to 1

Thalley reviewed (approved):
Minor suggestions, but is also OK as is

Thalley commented:
@jhedberg should this get merged for 3.7?

alwa-nordic commented on subsys/bluetooth/host/scan.c:
```diff
@@ -620,6 +620,24 @@ void bt_hci_le_adv_ext_report(struct net_buf *buf)
+
+			net_buf_reset(buf);
+
+			if (evt_type & BT_HCI_LE_ADV_EVT_TYPE_LEGACY) {
+				return;
+			}
+
+			/* Start discarding irrespective of the `more_to_come` flag. We
+			 * assume we may have lost a partial adv report in the truncated
+			 * data.
+			 */
+			reassembling_advertiser.state = FRAG_ADV_DISCARDING;
```
I think the early return may cause us to produce garbled advertisement reports.

>  Start discarding irrespective of the `more_to_come` flag. We assume we may have lost a partial adv report in the truncated data.

The above comment is there because we are in a while-loop over reports. By returning early, we may be skipping over one or more reports. This can lead to a bad reassembly.

A bad reassembly is not a huge problem. The application has to tolerate malicious advertisements anyway. But when we did discussed this last time, we decided to prefer dropping data.

jori-nordic replied:
can we come up with a test case for this somehow?

jori-nordic replied:
but in this particular case, it's the controller that's malicious, no? so likely not a problem for "normal" use-cases. This early-return is more for protecting the host (and its CPU / resources I guess). My opinion is that we should just shut down bluetooth when that happens. It indicates something is wrong, either HCI driver synchronization or that the controller is very buggy or compromised.

alwa-nordic replied:
Yes. This is only relevant for HCI protocol violation. Either the controller did bad, or there was data corruption during HCI transport. It's not an exposed attack surface.

ceolin commented:
> @jhedberg should this get merged for 3.7?

It gets to be merged :)
